### ISSUE TITLE
Higher peak memory with torch.compile

### ISSUE THREAD

**Issue by ad8e** [pytorch/pytorch#122512]
### 🐛 Describe the bug

If I fuse the backward with compiled forward+loss, there's a higher peak memory than if I separate the backward from compiled forward+loss. It looks like the logits aren't being cleared.

Fused forward+loss+backward:
```
@torch.compile
def fused_forward_and_loss_and_backward(input_ids, labels):
    logits = model.forward(input_ids)
    loss = F.cross_entropy(logits.view(-1, logits.shape[-1]).float(), labels.view(-1))
    # del logits # doesn't change peak memory
    loss.backward()
    return loss
...
# usage:
loss = fused_forward_and_loss_and_backward(batch_input_ids, gas_labels)
print("peak memory usage", torch.cuda.max_memory_allocated())
```
Results:
```
peak memory usage 22139266048 (first step)
peak memory usage 23943063552 (second step)
peak memory usage 23943063552 (third step)
```

Fused forward+loss, separated backward:

```
@torch.compile
def fused_forward_and_loss(input_ids, labels):
    logits = model.forward(input_ids)
    loss = F.cross_entropy(logits.view(-1, logits.shape[-1]).float(), labels.view(-1))
    # loss.backward() # no backward here!
    return loss
...
# usage:
loss = fused_forward_and_loss(batch_input_ids, gas_labels)
loss.backward()
print("peak memory usage", torch.cuda.max_memory_allocated())
```

Results:
```
peak memory usage 19991782400
peak memory usage 21795579392
peak memory usage 21795579392
```

Memory traces:
https://drive.google.com/file/d/18UywAfWmBDNJMzbCy44KVUy6qCug3cxX/view?usp=sharing
https://drive.google.com/file/d/1A0Cu9fAbJS1dbzBJvRmm-0U1ygsmtT9W/view?usp=sharing

### Versions

Collecting environment information...
PyTorch version: 2.2.0
Is debug build: False
CUDA used to build PyTorch: 12.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.19.17-coreweave-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A40
Nvidia driver version: 535.161.07
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          96
On-line CPU(s) list:             0-95
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7413 24-Core Processor
CPU family:                      25
Model:                           1
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
Stepping:                        1
Frequency boost:                 enabled
CPU max MHz:                     3630.8101
CPU min MHz:                     1500.0000
BogoMIPS:                        5299.98
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                  AMD-V
L1d cache:                       1.5 MiB (48 instances)
L1i cache:                       1.5 MiB (48 instances)
L2 cache:                        24 MiB (48 instances)
L3 cache:                        256 MiB (8 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-23,48-71
NUMA node1 CPU(s):               24-47,72-95
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.2.0
[pip3] torchaudio==2.2.0
[pip3] torchvision==0.17.0
[pip3] triton==2.2.0
[conda] Could not collect

**Comment by Chillee** [comment]
@ad8e Could you show the result of `TORCH_LOGS="aot_graphs"` for these?

**Comment by ad8e** [comment]
[unfused.txt](https://github.com/pytorch/pytorch/files/14763975/unfused.txt)
[fused.txt](https://github.com/pytorch/pytorch/files/14763977/fused.txt)

**Comment by Chillee** [comment]
I'm pretty sure what's going on is that today, Dynamo "graph breaks" when it encounters a `.backward()` call.

So, in the first version, it doesn't know that `logits` is no longer needed when it enters the backwards pass. So `logits` is kept alive for the duration of the backwards pass (which is why you see the increased peak memory).

In the second version, upon the function returning, all intermediates are freed, so we know they're no longer needed.

I think in theory adding `del` should help. @anijain2305 do you know whether we take that into account?

Also, to be clear, is the first version's memory usage on par with eager?

**Comment by ad8e** [comment]
Eager mode (no compile) has higher memory than both of the compiled runs (with fused backward and without fused backward).

With torch.compile, `del` did not change the peak memory; the number is the same as if no `del`.

`TORCH_LOGS="aot_graphs"`:
[compile on, del logits.txt](https://github.com/pytorch/pytorch/files/14764859/compile.on.del.logits.txt)
[compile on, no del.txt](https://github.com/pytorch/pytorch/files/14764862/compile.on.no.del.txt)
[compile off, del logits.txt](https://github.com/pytorch/pytorch/files/14764860/compile.off.del.logits.txt)

**Comment by Chillee** [comment]
Does `del` help eager mode's memory?

**Comment by Chillee** [comment]
Oh also, this PR might be relevant haha: https://github.com/pytorch/pytorch/pull/122658

**Comment by yifuwang** [comment]
> I'm pretty sure what's going on is that today, Dynamo "graph breaks" when it encounters a .backward() call.

> So, in the first version, it doesn't know that logits is no longer needed when it enters the backwards pass. So logits is kept alive for the duration of the backwards pass (which is why you see the increased peak memory).

@Chillee I have a small repro for this. Hope it helps:

```python
import pickle
import tempfile
from io import BytesIO

import torch


def graph(a):
    return a + 1, a + 2


# +-- Toggle this
# |
# v
@torch._dynamo.disable()
def graph_break(a):
    return a + 1


def resume(a, b):
    # Without graph break, a and b can get deallocated after `out_0 = a + b`;
    # With graph break, a and b doesn't get deallocated due to references
    # in the resumption function and its caller.
    out_0 = a + b
    out_1 = out_0 * 2
    out_2 = out_1 * 2
    return out_0, out_1, out_2


def fn(a):
    a, b = graph(a)
    a = graph_break(a)
    return resume(a, b)


a = torch.randn(10000, 10000, device="cuda", requires_grad=True)
compiled = torch.compile(fn, backend="eager")

torch.cuda.memory._record_memory_history(max_entries=100000)
for i in range(5):
    with torch.no_grad():
        compiled(a)

snapshot = torch.cuda.memory._snapshot()
with tempfile.NamedTemporaryFile() as f:
    f.write(BytesIO(pickle.dumps(snapshot)).getvalue())
torch.cuda.memory._record_memory_history(enabled=None)
```

Without graph break:
<img width="1563" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/d076f56d-7ff5-4769-98ab-dae4a94acd3a">

With graph break:
<img width="1565" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/e84987f2-1f3b-48c7-8ec0-5bd69a0a8c1c">

Bytecode with graph break:
```
ORIGINAL BYTECODE fn /home/yifu/microbench/resume_arg_2.py line 60
 61           0 LOAD_GLOBAL              0 (graph)
              2 LOAD_FAST                0 (a)
              4 CALL_FUNCTION            1
              6 UNPACK_SEQUENCE          2
              8 STORE_FAST               0 (a)
             10 STORE_FAST               1 (b)

 62          12 LOAD_GLOBAL              1 (graph_break)
             14 LOAD_FAST                0 (a)
             16 CALL_FUNCTION            1
             18 STORE_FAST               0 (a)

 63          20 LOAD_GLOBAL              2 (resume)
             22 LOAD_FAST                0 (a)
             24 LOAD_FAST                1 (b)
             26 CALL_FUNCTION            2
             28 RETURN_VALUE


MODIFIED BYTECODE fn /home/yifu/microbench/resume_arg_2.py line 60
 60           0 LOAD_GLOBAL              3 (__compiled_fn_0)
              2 LOAD_FAST                0 (a)
              4 CALL_FUNCTION            1
              6 STORE_FAST               2 (graph_out_0)
              8 LOAD_GLOBAL              1 (graph_break)
             10 LOAD_FAST                2 (graph_out_0)
             12 LOAD_CONST               1 (0)
             14 BINARY_SUBSCR
             16 LOAD_FAST                2 (graph_out_0)
             18 LOAD_CONST               2 (1)
             20 BINARY_SUBSCR
             22 STORE_FAST               1 (b)
             24 DELETE_FAST              2 (graph_out_0)

 62          26 CALL_FUNCTION            1
             28 LOAD_GLOBAL              4 (__resume_at_18_1)
             30 ROT_TWO
             32 LOAD_FAST                1 (b)
             34 CALL_FUNCTION            2
             36 RETURN_VALUE

ORIGINAL BYTECODE torch_dynamo_resume_in_fn_at_62 /home/yifu/microbench/resume_arg_2.py line 62
 62           0 LOAD_FAST                0 (___stack0)
              2 JUMP_ABSOLUTE           11 (to 22)
              4 LOAD_GLOBAL              0 (graph)
              6 LOAD_FAST                2 (a)
              8 CALL_FUNCTION            1
             10 UNPACK_SEQUENCE          2
             12 STORE_FAST               2 (a)
             14 STORE_FAST               1 (b)
             16 LOAD_GLOBAL              1 (graph_break)
             18 LOAD_FAST                2 (a)
             20 CALL_FUNCTION            1
        >>   22 STORE_FAST               2 (a)

 63          24 LOAD_GLOBAL              2 (resume)
             26 LOAD_FAST                2 (a)
             28 LOAD_FAST                1 (b)
             30 CALL_FUNCTION            2
             32 RETURN_VALUE


MODIFIED BYTECODE torch_dynamo_resume_in_fn_at_62 /home/yifu/microbench/resume_arg_2.py line 62
 62           0 LOAD_GLOBAL              3 (__compiled_fn_2)
              2 LOAD_FAST                0 (___stack0)
              4 LOAD_FAST                1 (b)
              6 CALL_FUNCTION            2
              8 STORE_FAST               3 (graph_out_0)
             10 LOAD_FAST                3 (graph_out_0)
             12 LOAD_CONST               1 (0)
             14 BINARY_SUBSCR
             16 LOAD_FAST                3 (graph_out_0)
             18 LOAD_CONST               2 (1)
             20 BINARY_SUBSCR
             22 LOAD_FAST                3 (graph_out_0)
             24 LOAD_CONST               3 (2)
             26 BINARY_SUBSCR
             28 BUILD_TUPLE              3
             30 DELETE_FAST              3 (graph_out_0)
             32 RETURN_VALUE
```

**Comment by Chillee** [comment]
@yifuwang Does your PR solve the problem?

**Comment by yifuwang** [comment]
@Chillee no. The PR only fixes the case where a graph output is last used in a graph break, but is being kept alive throughout the subsequent graph. Though the PR removes a reference that contributes to this problem.

**Comment by bhack** [comment]
Same problem here with nightly. In eager with `del Q, QK, K` is consuming less memory that with `@torch.compile`
```python
    def forward(self, Q, K, V):
        """
        :param Q: A 3d tensor with shape of [T_q, bs, C_q]
        :param K: A 3d tensor with shape of [T_k, bs, C_k]
        :param V: A 3d tensor with shape of [T_v, bs, C_v]
        """
       #print_cuda_memory()

        num_head = self.num_head
        hidden_dim = self.hidden_dim

        bs = Q.size()[1]

        # Linear projections
        if self.use_linear:
            Q = self.linear_Q(Q)
            K = self.linear_K(K)
            V = self.linear_V(V)

        # Scale
        Q = Q / self.T

        if not self.training and self.max_mem_len_ratio > 0:
            mem_len_ratio = float(K.size(0)) / Q.size(0)
            if mem_len_ratio > self.max_mem_len_ratio:
                scaling_ratio = math.log(mem_len_ratio) / math.log(
                    self.max_mem_len_ratio)
                Q = Q * scaling_ratio

        # Multi-head
        Q = Q.view(-1, bs, num_head, self.d_att).permute(1, 2, 0, 3)
        K = K.view(-1, bs, num_head, self.d_att).permute(1, 2, 3, 0)
        V = V.view(-1, bs, num_head, hidden_dim).permute(1, 2, 0, 3)

        # Multiplication
        QK = multiply_by_ychunks(Q, K, self.qk_chunks)
        if self.use_dis:
            QK = 2 * QK - K.pow(2).sum(dim=-2, keepdim=True)
        # Activation
        if not self.training and self.top_k != -1 and self.top_k < QK.size()[-1]:
            if self.top_k > 1:
                k = int(self.top_k)
            elif self.top_k > 0 and self.top_k < 1:
                k = int(QK.size()[-1] * self.top_k)
            elif self.top_k > -1 and self.top_k < 0:
                k = int(QK.size()[-2] * (-self.top_k))
            else:
                raise NotImplementedError
            top_QK, indices = torch.topk(QK, k=k, dim=-1)
            top_attn = torch.softmax(top_QK, dim=-1)
            attn = torch.zeros_like(QK).scatter_(-1, indices, top_attn)
        else:
            attn = torch.softmax(QK, dim=-1)
        del Q, QK, K
        # Dropouts
        attn = self.dropout(attn)
        #print_cuda_memory()
        # Weighted sum
        outputs = multiply_by_xchunks(attn, V,
                                      self.qk_chunks).permute(2, 0, 1, 3)
        # Restore shape
        outputs = outputs.reshape(-1, bs, self.d_model)

        outputs = self.projection(outputs)

        return outputs, attn
```

**Comment by masnesral** [comment]
[old issue scrubbing] Looks like this is still relevant; I don't think @Chillee is gonna work on it though, so removing. I'm also not sure this is inductor, so being presumptuous and removing for better tracking.
